### ISSUE TITLE
Reproducing the BASE model.

### ISSUE THREAD

**Issue by cxxgtxy** [LTH14/mar#48]
Thanks for releasing the code.
I use the code to reproduce the result of the base model. However, the FID is much higher than your result.
I carefully check the issue of https://github.com/LTH14/mar/issues/16. 

I enable online eval and calculate the metric (without CFG) every 40 epochs. It seems that FID changed very slowly after 200 epches.

[16:45:18.199366] FID: 564.4450, Inception Score: 1.0000
[22:45:44.034328] FID: 344.6018, Inception Score: 1.5078
[04:46:30.035496] FID: 200.7698, Inception Score: 4.7345
[10:45:39.495206] FID: 41.3462, Inception Score: 77.7484
[16:45:47.695665] FID: 41.1176, Inception Score: 78.7093
[22:45:36.932066] FID: 41.4553, Inception Score: 78.6408
[04:45:21.494827] FID: 40.7288, Inception Score: 78.8541
[10:45:58.329206] FID: 40.4982, Inception Score: 79.7356
[16:47:08.393834] FID: 41.1990, Inception Score: 80.4728
[22:46:10.148548] FID: 40.8915, Inception Score: 80.2271
[04:45:07.046069] FID: 40.7950, Inception Score: 79.7204
[10:48:45.949598] FID: 41.0686, Inception Score: 81.3697
[16:53:52.302475] FID: 40.5346, Inception Score: 79.9995
[22:48:39.635982] FID: 40.7077, Inception Score: 81.2267
[04:51:48.963065] FID: 40.6648, Inception Score: 81.0701
[10:56:57.583088] FID: 40.8566, Inception Score: 79.3724
[17:01:11.069810] FID: 40.8156, Inception Score: 80.9139
[23:04:09.087249] FID: 40.7860, Inception Score: 79.9918
[05:10:56.585507] FID: 40.2217, Inception Score: 79.5285

When I test the  model with FID: 40.2217, Inception Score: 79.5285 (without CFG), its CFG result is  FID: 5.0722, Inception Score: 164.7981.

I wonder whether you have noticed the FID  trend (wo or w CFG) during the training. If you do, what's your trend? Thanks

**Comment by LTH14** [comment]
This is weird. Here is our online eval curve of MAR-B model:
without cfg (cfg=1.0), 64 steps:
<img width="984" alt="image" src="https://github.com/user-attachments/assets/44b4555a-8ad0-4e48-b8ce-bacdca544f80">

with cfg=3.0, 64 steps
<img width="984" alt="image" src="https://github.com/user-attachments/assets/2d3c6543-b5b7-4023-9d2c-f7d73dcc23a5">

It is normal for the FID to drop significantly at 160-200 epochs -- this is because we use ema=0.9999. However, ~40 FID is too high. Did you exactly follow the training script provided, with only mar_large changed to mar_base?

**Comment by cxxgtxy** [comment]
Yes. I exactly follow the training script. Here is my training script. 4 Nodes
```shell
model_name=mar_base
torchrun --nproc_per_node=$gpu_per_pod --nnodes=$WORLD_SIZE --node_rank=$RANK --master_addr=${MASTER_ADDR} --master_port=${MASTER_PORT} \
 main_mar.py --img_size 256 --vae_path pretrained_models/vae/kl16.ckpt --vae_embed_dim 16 --vae_stride 16 \
 --patch_size 1 --model $model_name --diffloss_d 3 --diffloss_w 1024 --epochs 400 --warmup_epochs 100 --batch_size 64 \
 --blr 1.0e-4 --diffusion_batch_mul 4 --data_path $imagenet_dir --output_dir ${model_name} --resume ${model_name} \
 --use_cached --cached_path $imagenet_dir --online_eval --save_last_freq 40
```

**Comment by LTH14** [comment]
Can you try diffloss_d 6 and diffloss_w 1024? I never tried a MAR-B with depth=3. MAR-L with depth=3 and width 1024 works fine though.

**Comment by cxxgtxy** [comment]
Sorry.  This is the script for 800 epochs training. The former one is 400 epochs. 

```
#torchrun --nproc_per_node=$gpu_per_pod --nnodes=$WORLD_SIZE --node_rank=$RANK --master_addr=${MASTER_ADDR} --master_port=${MASTER_PORT} \
#main_mar.py --img_size 256 --vae_path pretrained_models/vae/kl16.ckpt --vae_embed_dim 16 --vae_stride 16 \
#--patch_size 1 --model $model_name --diffloss_d 6 --diffloss_w 1024 --epochs 800 --warmup_epochs 100 --batch_size 64 \
#--blr 1.0e-4 --diffusion_batch_mul 4 --data_path $imagenet_dir --output_dir ${model_name}_e800 --save_dir "/var/data/" --resume ${model_name}_e800 \
#--use_cached --cached_path $imagenet_dir --online_eval --save_last_freq 40
```

**Comment by LTH14** [comment]
This is weird -- this should exactly reproduce the FID curve. Have you tried mar_large? Does it give the same behavior?

**Comment by cxxgtxy** [comment]
I enable use_cached to speed up the training process.
And the weight is from https://huggingface.co/jadechoghari/mar because I cannot access the VAE cpt in your repo.
It should not affect the result.

**Comment by cxxgtxy** [comment]
> This is weird -- this should exactly reproduce the FID curve. Have you tried mar_large? Does it give the same behavior?

I haven't tried the Large model because of limited resource.

**Comment by cxxgtxy** [comment]
Add the loss stats of last 20 epochs for the base with 800 epochs.

[08:08:44.635347] Averaged stats: lr: 0.000800  loss: 0.3260 (0.3178)
[08:17:31.752362] Averaged stats: lr: 0.000800  loss: 0.3219 (0.3177)
[08:26:27.177488] Averaged stats: lr: 0.000800  loss: 0.3141 (0.3180)
[08:35:21.123790] Averaged stats: lr: 0.000800  loss: 0.3076 (0.3183)
[08:44:28.315774] Averaged stats: lr: 0.000800  loss: 0.2987 (0.3180)
[08:53:29.075227] Averaged stats: lr: 0.000800  loss: 0.3208 (0.3183)
[09:02:23.759449] Averaged stats: lr: 0.000800  loss: 0.3051 (0.3178)
[09:11:17.584541] Averaged stats: lr: 0.000800  loss: 0.3120 (0.3183)
[09:20:10.872805] Averaged stats: lr: 0.000800  loss: 0.3081 (0.3178)
[09:28:59.632466] Averaged stats: lr: 0.000800  loss: 0.3154 (0.3182)
[09:37:44.397939] Averaged stats: lr: 0.000800  loss: 0.3063 (0.3187)
[09:46:30.271498] Averaged stats: lr: 0.000800  loss: 0.3031 (0.3177)
[09:55:27.214555] Averaged stats: lr: 0.000800  loss: 0.2941 (0.3179)
[10:04:18.622682] Averaged stats: lr: 0.000800  loss: 0.3001 (0.3179)
[10:13:21.334477] Averaged stats: lr: 0.000800  loss: 0.3038 (0.3182)
[10:22:15.467774] Averaged stats: lr: 0.000800  loss: 0.3026 (0.3177)
[10:31:03.857089] Averaged stats: lr: 0.000800  loss: 0.3148 (0.3179)
[10:39:53.369399] Averaged stats: lr: 0.000800  loss: 0.3045 (0.3179)
[10:48:48.786469] Averaged stats: lr: 0.000800  loss: 0.3018 (0.3179)
[10:57:59.641412] Averaged stats: lr: 0.000800  loss: 0.3289 (0.3179)

**Comment by LTH14** [comment]
Here's the loss curve
<img width="507" alt="image" src="https://github.com/user-attachments/assets/d98fdc45-38d6-46e2-bb09-dd5ebaf1ebd3">

**Comment by cxxgtxy** [comment]
Can you provide the VAE cpt?  I check whether the vae from   https://huggingface.co/jadechoghari/mar is the same as yours?

**Comment by LTH14** [comment]
The VAE checkpoint can be downloaded using ```python util/download.py```. You can also download it manually using [this link](https://www.dropbox.com/scl/fi/hhmuvaiacrarfg28qxhwz/kl16.ckpt?rlkey=l44xipsezc8atcffdp4q7mwmh&dl=0)

**Comment by cxxgtxy** [comment]
> The VAE checkpoint can be downloaded using `python util/download.py`. You can also download it manually using [this link](https://www.dropbox.com/scl/fi/hhmuvaiacrarfg28qxhwz/kl16.ckpt?rlkey=l44xipsezc8atcffdp4q7mwmh&dl=0)

I know, but I am from a nation where these links are unavailable.

**Comment by cxxgtxy** [comment]
Thanks. I find the problem.

**Comment by shaochenze** [comment]
Hi @LTH14 , I am reproducing the default MAR setting (MAR-large, 3 MLP blocks, 400 epochs). I observed a significant drop of FID at epoch=160, but after that the FID shows no further decrease. Here's the FID curve during training (evaluated at temperature=1). Is this FID curve expected, or something has gone wrong? Also, would changing the temperature to 0.95 help see the right FID curve? 
<img width="1091" alt="fid" src="https://github.com/user-attachments/assets/986e9122-10a6-4360-8b6f-381a3ba34c97">
Thanks for your help!

**Comment by LTH14** [comment]
Yes this FID curve is expected -- the significant drop is mainly because of the ema we use. Changing the temperature to 0.95 would give you a better FID value, but the trend would be similar (i.e., a significant drop at 160 epochs and a slight decrease after 160 epochs).

**Comment by shaochenze** [comment]
Thank you for the clarification @LTH14 ! I'm still a bit confused. In my training, the best FID score was obtained at epoch 160. Contrary to expecting a slight decrease in FID after epoch=160, I observed an increase in the FID score instead.

**Comment by LTH14** [comment]
Yes this slight increase is because of the unoptimized temperature. I don't have a curve for the 0.95 temperature, but here is the curve for cfg=3.0 and temperature=1.0:
<img width="986" alt="image" src="https://github.com/user-attachments/assets/08b7e283-5d74-4ba2-96d0-f39f897e9eb8">

**Comment by shaochenze** [comment]
Thank you!

**Comment by Doraemonzzz** [comment]
I also encountered issues when reproducing the base model. Here are my training curve and training script. (Currently, 150 epochs have been trained.)
![image](https://github.com/user-attachments/assets/03b56023-70d9-4d16-8b71-1208b492cc5a)

![image](https://github.com/user-attachments/assets/cb8c20b2-d831-40ce-85a7-870fa4ae9b80)

```
main_mar.py \
    --img_size 256 --vae_path ${ckpt} --vae_embed_dim 16 --vae_stride 16 \
    --patch_size 1 --model $model_name --diffloss_d 6 --diffloss_w 1024 --epochs 400 --warmup_epochs 100 --batch_size ${batch_size} \
    --blr 1.0e-4 --diffusion_batch_mul 4 --output_dir ${model_name} --resume ${model_name} \
    --use_cached --cached_path $imagenet_dir --use_cached --online_eval --save_last_freq 40
```

**Comment by LTH14** [comment]
You should wait until the 160 epoch because we are using ema.

**Comment by Doraemonzzz** [comment]
> You should wait until the 160 epoch because we are using ema.

Thank you for your response. Everything seems to be normal now, and here is the subsequent FID. On the other hand, I would like to ask you to confirm the impact of EMA on this task. For example, if the EMA is not used, would we still achieve a relatively low FID? Do you have any insights on why the FID suddenly dropped from several hundred to around 5?

<img width="1688" alt="image" src="https://github.com/user-attachments/assets/0e21e257-c77e-4aec-951c-192747e61c4b">

**Comment by LTH14** [comment]
Without ema the FID should be slightly higher at the end, but still <4. Ema causes a delay in the model parameters, so it takes a longer time to get a reasonable model (160 epochs)

**Comment by Doraemonzzz** [comment]
After retraining for 400 epochs, the FID (without CFG) of the base model is 5.26269, and the FID (without CFG) of the large model is 4.23104. Is this result normal?

**Comment by LTH14** [comment]
You should change the temperature to 0.95 to test the FID without CFG.

**Comment by Doraemonzzz** [comment]
The updated latest evaluation results are as follows:

model | fid | epoch | cfg | temperature
-- | -- | -- | -- | --
Mar base | 5.26269 | 400 | 1 | 1
Mar base | 3.8846 | 400 | 1 | 0.95
Mar large | 4.23104 | 400 | 1 | 1
Mar large | 3.1612 | 400 | 1 | 0.95

**Comment by LTH14** [comment]
This is worse than expected. Did you follow the training script in the readme exactly? What about the FID with cfg=3.0?

**Comment by shaochenze** [comment]
> Without ema the FID should be slightly higher at the end, but still <4. Ema causes a delay in the model parameters, so it takes a longer time to get a reasonable model (160 epochs)

Hi @LTH14 , I tried not to use EMA for evaluation, and the difference is very significant. The FID is 3.88 w/ EMA and 10.24 w/o EMA. Have you ever encountered such a situation?

**Comment by LTH14** [comment]
This is expected. EMA is very important in the FID number.

**Comment by shaochenze** [comment]
Thank you! This is my first time working on image-related tasks, so I don't quite understand why EMA is so important. Is EMA necessary for all visual models, or is it only used in diffusion models? In other modalities, such as text, EMA doesn't seem to play such a crucial role. Could you share some insights?

**Comment by LTH14** [comment]
EMA is a quite standard trick used in many generative models, date back to the GAN era. You can check this paper for a more comprehensive evaluation of EMA https://arxiv.org/pdf/2312.02696
